File: 0983-minimum-cost-for-tickets/0983-minimum-cost-for-tickets.cpp
class Solution {
public:
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        int minCost[366];
        minCost[0] = 0;
        
        unordered_map<int,bool> daysIncluded;
        for(auto day : days)
            daysIncluded[day]=true;
        
        for(int i=1;i<=365;i++)
        {
            if(!daysIncluded[i])
                minCost[i]=minCost[i-1];
            else
            {
                int _1DayPass = minCost[i-1] + costs[0];
                int _7DayPass = minCost[max(0,i-7)] + costs[1];
                int _30DayPass = minCost[max(0,i-30)] + costs[2];

                minCost[i] = min(_1DayPass,min(_7DayPass,_30DayPass));
            }
        }
        return minCost[365];
    }  
};

File: 0983-minimum-cost-for-tickets/0983-minimum-cost-for-tickets.py
class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        prev = 0
        leng = len(days)
        minarr = [float('inf')]*leng
        mc = min(costs)
        
        for i in range(leng):

            minarr[i] = min(minarr[i],prev+mc)
            day = days[i]
            
            #7 day
            limit = min(day+6,days[-1])
            reach = bisect.bisect_right(days,limit)-1
            minarr[reach] = min(prev+costs[1],minarr[reach])

            #30 day
            curr = i
            limit = min(day+29,days[-1])
            reach = bisect.bisect_right(days,limit)-1
            minarr[reach] = min(prev+costs[2],minarr[reach])

            prev = minarr[i]

        return minarr[-1]